pkg/arm/authorizer.go

```diff
@@ -0,0 +1,161 @@
+			return nil, err
+		}
+		settings.environment = val
+	}
+
+	if settings.resource == "" {
+		settings.resource = strings.TrimSuffix(settings.environment.KeyVaultEndpoint, "/")
+	}
+
+	// 1. Client credentials
+	if (settings.clientSecret != "") || settings.authMethod == "clientcredentials" {
+		config := auth.NewClientCredentialsConfig(settings.clientID, settings.clientSecret, settings.tenantID)
+		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
+		config.Resource = settings.resource
+		return config.Authorizer()
+	}
+
+	// 2. Client Certificate
+	if (settings.certificatePath != "") || settings.authMethod == "clientcertificate" {
+		config := auth.NewClientCertificateConfig(settings.certificatePath, settings.certificatePassword, settings.clientID, settings.tenantID)
+		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
+		config.Resource = settings.resource
+		return config.Authorizer()
+	}
+
+	// 3. Username Password
+	if (settings.username != "" && settings.password != "") || settings.authMethod == "usernamepassword" {
+		config := auth.NewUsernamePasswordConfig(settings.username, settings.password, settings.clientID, settings.tenantID)
+		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
+		config.Resource = settings.resource
+		return config.Authorizer()
+	}
+
+	// 4. MSI
+	if settings.authMethod == "msi" {
+		config := auth.NewMSIConfig()
+		config.Resource = settings.resource
+		config.ClientID = settings.clientID
+		return config.Authorizer()
+	}
+
+	// TODO: decide how to handle prompt on stdout, caching, etc.
+	// 5. Device Code
+	if settings.authMethod == "devicecode" {
+		// TODO: This will be required on every execution. Consider caching.
+		config := auth.NewDeviceFlowConfig(settings.clientID, settings.tenantID)
+		return config.Authorizer()
+	}
+
+	// 5. CLI
+	return auth.NewAuthorizerFromCLIWithResource(settings.resource)
+
+}
+
+func NewDeviceCodeAuthorizerWithCache() (autorest.Authorizer, error) {
+	// using only the settings relevent to this authorizer
+	settings := struct {
+		tenantID string
+		clientID string
+	}{
+		tenantID: os.Getenv("AZURE_TENANT_ID"),
+		clientID: os.Getenv("AZURE_CLIENT_ID"),
+	}
+
+	// TODO: remove hard-coded cache location
+	dir, err := homedir.Dir()
+	if err != nil {
+		return nil, err
+	}
+	tokenCachePath := filepath.Join(dir, ".a3e")
+
+	// load token from cache if available
+	token, err := adal.LoadToken(tokenCachePath)
+	config := auth.NewDeviceFlowConfig(settings.clientID, settings.tenantID)
+
+	if err != nil {
+		// authenticate via devicecode flow
+		config := auth.NewDeviceFlowConfig(settings.clientID, settings.tenantID)
+		spt, err := config.ServicePrincipalToken()
+		if err != nil {
+			return nil, err
+		}
+		tmp := spt.Token()
+		token = &tmp
+		err = adal.SaveToken(tokenCachePath, 0600, *token)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		oauthConfig, err := adal.NewOAuthConfig(config.AADEndpoint, config.TenantID)
```

Comment:
I feel that there ought to have been a simpler way to refresh the token loaded when from cache. Will look into this sometime.